Allow configuring the theme for syntax highlighting

Looking into a way to make the syntax highlighting theme adjustable from the config file.
For example it would be great if I could write something like this in the `.cobalt.yml`:

```yaml
theme: base16-ocean.dark
```

I have a rough prototype here: https://github.com/cobalt-org/cobalt.rs/compare/master...mre:config-theme

Unfortunately I need some more guidance for the next steps.

The part around `CodeBlock`s is not working yet.
What I wanted to do is inject the config as a parameter into the codeblock.
But the codeblock needs to implement `Renderable` from Liquid. 
(https://github.com/cobalt-org/cobalt.rs/compare/master...mre:config-theme#diff-b9237993dc7f8fb4f95923e1c2a64499R46)

The error message I get is:
```Rust
# cargo build --features="syntax-highlight"
   Compiling cobalt-bin v0.6.1 (file:///.../cobalt.rs)
error[E0281]: type mismatch: the type `fn(&str, &[liquid::Token], std::vec::Vec<liquid::lexer::Element>, &liquid::LiquidOptions, &config::Config) -> std::result::Result<Box<liquid::Renderable>, liquid::Error> {syntax_highlight::initialize_codeblock}` implements the trait `
for<'r, 'r, 'r, 'r> std::ops::Fn<(&'r str, &'r [liquid::Token], std::vec::Vec<liquid::lexer::Element>, &'r liquid::LiquidOptions, &'r config::Config)>`, but the trait `for<'r, 'r, 'r> std::ops::Fn<(&'r str, &'r [liquid::Token], std::vec::Vec<liquid::lexer::Element>, &'r li
quid::LiquidOptions)>` is required (expected a tuple with 4 elements, found one with 5 elements)
   --> src/document.rs:272:56
    |
272 |         options.blocks.insert("highlight".to_string(), Box::new(initialize_codeblock));
    |                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: required for the cast to the object type `for<'r, 'r, 'r> std::ops::Fn(&'r str, &'r [liquid::Token], std::vec::Vec<liquid::lexer::Element>, &'r liquid::LiquidOptions) -> std::result::Result<Box<liquid::Renderable + 'static>, liquid::Error> + 'static`
```

That should be the last roadblock. Maybe you can give me a hint on what I could try.

--- a/src/cobalt.rs
+++ b/src/cobalt.rs
@@ -183,8 +183,12 @@ pub fn build(config: &Config) -> Result<()> {
 
         let mut context = post.get_render_context(&simple_posts_data);
 
-        post.render_excerpt(&mut context, source, config)?;
-        let post_html = post.render(&mut context, source, &layouts, &mut layouts_cache, config)?;
+        post.render_excerpt(&mut context, source, &config.syntax_theme)?;
+        let post_html = post.render(&mut context,
+                                    source,
+                                    &layouts,
+                                    &mut layouts_cache,
+                                    &config.syntax_theme)?;
         create_document_file(post_html, &post.file_path, dest)?;
     }
 
@@ -224,7 +228,11 @@ pub fn build(config: &Config) -> Result<()> {
         }
 
         let mut context = doc.get_render_context(&posts_data);
-        let doc_html = doc.render(&mut context, source, &layouts, &mut layouts_cache, config)?;
+        let doc_html = doc.render(&mut context,
+                                  source,
+                                  &layouts,
+                                  &mut layouts_cache,
+                                  &config.syntax_theme)?;
         create_document_file(doc_html, doc.file_path, dest)?;
     }
 

--- a/src/cobalt.rs
+++ b/src/cobalt.rs
@@ -183,8 +183,8 @@ pub fn build(config: &Config) -> Result<()> {
 
         let mut context = post.get_render_context(&simple_posts_data);
 
-        post.render_excerpt(&mut context, source)?;
-        let post_html = post.render(&mut context, source, &layouts, &mut layouts_cache)?;
+        post.render_excerpt(&mut context, source, config)?;
+        let post_html = post.render(&mut context, source, &layouts, &mut layouts_cache, config)?;
         create_document_file(post_html, &post.file_path, dest)?;
     }
 
@@ -224,7 +224,7 @@ pub fn build(config: &Config) -> Result<()> {
         }
 
         let mut context = doc.get_render_context(&posts_data);
-        let doc_html = doc.render(&mut context, source, &layouts, &mut layouts_cache)?;
+        let doc_html = doc.render(&mut context, source, &layouts, &mut layouts_cache, config)?;
         create_document_file(doc_html, doc.file_path, dest)?;
     }
 

--- a/src/config.rs
+++ b/src/config.rs
@@ -39,6 +39,7 @@ pub struct Config {
     pub ignore: Vec<String>,
     pub excerpt_separator: String,
     pub dump: Vec<Dump>,
+    pub theme: String,
 }
 
 impl Default for Config {
@@ -61,6 +62,7 @@ impl Default for Config {
             ignore: vec![],
             excerpt_separator: "\n\n".to_owned(),
             dump: vec![],
+            theme: "base16-ocean.dark".to_owned(),
         }
     }
 }
@@ -139,6 +141,10 @@ impl Config {
 
         if let Some(excerpt_separator) = yaml["excerpt_separator"].as_str() {
             config.excerpt_separator = excerpt_separator.to_owned();
+        };
+
+        if let Some(theme) = yaml["theme"].as_str() {
+            config.theme = theme.to_owned();
         };
 
         Ok(config)

--- a/src/config.rs
+++ b/src/config.rs
@@ -5,6 +5,9 @@ use std::io::Read;
 use error::Result;
 use yaml_rust::YamlLoader;
 
+#[cfg(all(feature="syntax-highlight", not(windows)))]
+use syntax_highlight::has_syntax_theme;
+
 arg_enum! {
     #[derive(Debug, PartialEq, Copy, Clone)]
     pub enum Dump {
@@ -143,9 +146,17 @@ impl Config {
             config.excerpt_separator = excerpt_separator.to_owned();
         };
 
-        if let Some(theme) = yaml["syntax-highlight"]["theme"].as_str() {
-            config.syntax_theme = theme.to_owned();
-        };
+        #[cfg(all(feature="syntax-highlight", not(windows)))]
+        {
+            if let Some(theme) = yaml["syntax-highlight"]["theme"].as_str() {
+                if has_syntax_theme(theme) {
+                    config.syntax_theme = theme.to_owned();
+                } else {
+                    warn!("Theme named '{}' is not found.  Using default 'base16-ocean.dark'.",
+                          theme);
+                }
+            };
+        }
 
         Ok(config)
     }

--- a/src/config.rs
+++ b/src/config.rs
@@ -39,7 +39,7 @@ pub struct Config {
     pub ignore: Vec<String>,
     pub excerpt_separator: String,
     pub dump: Vec<Dump>,
-    pub theme: String,
+    pub syntax_theme: String,
 }
 
 impl Default for Config {
@@ -62,7 +62,7 @@ impl Default for Config {
             ignore: vec![],
             excerpt_separator: "\n\n".to_owned(),
             dump: vec![],
-            theme: "base16-ocean.dark".to_owned(),
+            syntax_theme: "base16-ocean.dark".to_owned(),
         }
     }
 }
@@ -143,8 +143,8 @@ impl Config {
             config.excerpt_separator = excerpt_separator.to_owned();
         };
 
-        if let Some(theme) = yaml["theme"].as_str() {
-            config.theme = theme.to_owned();
+        if let Some(theme) = yaml["syntax-highlight"]["theme"].as_str() {
+            config.syntax_theme = theme.to_owned();
         };
 
         Ok(config)

--- a/src/config.rs
+++ b/src/config.rs
@@ -152,7 +152,7 @@ impl Config {
                 if has_syntax_theme(theme) {
                     config.syntax_theme = theme.to_owned();
                 } else {
-                    warn!("Theme named '{}' is not found.  Using default 'base16-ocean.dark'.",
+                    warn!("Theme named '{}' is not found.  Using default 'base16-ocean.dark'.  Run `cobalt list-syntax-themes` to see available themes.",
                           theme);
                 }
             };

--- a/src/config.rs
+++ b/src/config.rs
@@ -5,7 +5,7 @@ use std::io::Read;
 use error::Result;
 use yaml_rust::YamlLoader;
 
-#[cfg(all(feature="syntax-highlight", not(windows)))]
+#[cfg(all(feature="syntax-highlight"))]
 use syntax_highlight::has_syntax_theme;
 
 arg_enum! {
@@ -146,14 +146,26 @@ impl Config {
             config.excerpt_separator = excerpt_separator.to_owned();
         };
 
-        #[cfg(all(feature="syntax-highlight", not(windows)))]
+        #[cfg(all(feature="syntax-highlight"))]
         {
+            use error::{Error, ErrorKind};
+
             if let Some(theme) = yaml["syntax-highlight"]["theme"].as_str() {
-                if has_syntax_theme(theme) {
-                    config.syntax_theme = theme.to_owned();
-                } else {
-                    warn!("Theme named '{}' is not found.  Using default 'base16-ocean.dark'.  Run `cobalt list-syntax-themes` to see available themes.",
-                          theme);
+                match has_syntax_theme(theme) {
+                    Ok(true) => config.syntax_theme = theme.to_owned(),
+                    Ok(false) => {
+                        info!(concat!("Theme named '{}' is not found.",
+                                      "  Using default 'base16-ocean.dark'.",
+                                      "  Run `cobalt list-syntax-themes` to see available themes."),
+                              theme)
+                    }
+                    Err(Error(ErrorKind::UnsupportedPlatform(..), _)) => {
+                        info!(concat!("Theme named '{}' ignored.",
+                                      "  This build of cobalt does not include support",
+                                      " for syntax highlighting."),
+                              theme)
+                    }
+                    Err(_) => unreachable!(),
                 }
             };
         }

--- a/src/document.rs
+++ b/src/document.rs
@@ -13,7 +13,6 @@ use jsonfeed::Item;
 use jsonfeed::Content;
 use serde_yaml;
 use itertools;
-use config::Config;
 
 #[cfg(all(feature="syntax-highlight", not(windows)))]
 use syntax_highlight::{initialize_codeblock, decorate_markdown};
@@ -384,13 +383,14 @@ impl Document {
                    content: &str,
                    context: &mut Context,
                    source: &Path,
-                   config: &Config)
+                   syntax_theme: &str)
                    -> Result<String> {
-        let theme_name = config.syntax_theme.clone();
         let mut options = LiquidOptions::default();
         options.template_repository = Box::new(LocalTemplateRepository::new(source.to_owned()));
-        let highlight: Box<liquid::Block> =
-            Box::new(move |_, args, tokens, _| initialize_codeblock(args, tokens, &theme_name));
+        let highlight: Box<liquid::Block> = {
+            let syntax_theme = syntax_theme.to_owned();
+            Box::new(move |_, args, tokens, _| initialize_codeblock(args, tokens, &syntax_theme))
+        };
         options.blocks.insert("highlight".to_string(), highlight);
         let template = try!(liquid::parse(content, options));
         let html = try!(template.render(context)).unwrap_or_default();
@@ -401,7 +401,7 @@ impl Document {
                 let mut buf = String::new();
                 let parser = cmark::Parser::new(&html);
                 #[cfg(feature="syntax-highlight")]
-                cmark::html::push_html(&mut buf, decorate_markdown(parser, config));
+                cmark::html::push_html(&mut buf, decorate_markdown(parser, syntax_theme));
                 #[cfg(not(feature="syntax-highlight"))]
                 cmark::html::push_html(&mut buf, parser);
                 buf
@@ -414,7 +414,7 @@ impl Document {
                    content: &str,
                    context: &mut Context,
                    source: &Path,
-                   _: &Config)
+                   _syntax_theme: &str)
                    -> Result<String> {
         let mut options = LiquidOptions::default();
         options.template_repository = Box::new(LocalTemplateRepository::new(source.to_owned()));
@@ -455,7 +455,7 @@ impl Document {
     pub fn render_excerpt(&mut self,
                           context: &mut Context,
                           source: &Path,
-                          config: &Config)
+                          syntax_theme: &str)
                           -> Result<()> {
         let excerpt_html = {
             let excerpt_attr = self.attributes
@@ -465,14 +465,14 @@ impl Document {
             let excerpt_separator = &self.front.excerpt_separator;
 
             if let Some(excerpt_str) = excerpt_attr {
-                try!(self.render_html(excerpt_str, context, source, config))
+                try!(self.render_html(excerpt_str, context, source, syntax_theme))
             } else if excerpt_separator.is_empty() {
-                try!(self.render_html("", context, source, config))
+                try!(self.render_html("", context, source, syntax_theme))
             } else {
                 try!(self.render_html(&self.extract_markdown_references(excerpt_separator),
                                       context,
                                       source,
-                                      config))
+                                      syntax_theme))
             }
         };
 
@@ -495,9 +495,9 @@ impl Document {
                   source: &Path,
                   layouts_dir: &Path,
                   layouts_cache: &mut HashMap<String, String>,
-                  config: &Config)
+                  syntax_theme: &str)
                   -> Result<String> {
-        let content_html = try!(self.render_html(&self.content, context, source, config));
+        let content_html = try!(self.render_html(&self.content, context, source, syntax_theme));
         self.attributes
             .insert("content".to_owned(), Value::Str(content_html.clone()));
         context.set_val("content", Value::Str(content_html.clone()));

--- a/src/document.rs
+++ b/src/document.rs
@@ -386,9 +386,11 @@ impl Document {
                    source: &Path,
                    config: &Config)
                    -> Result<String> {
+        let theme_name = config.syntax_theme.clone();
         let mut options = LiquidOptions::default();
         options.template_repository = Box::new(LocalTemplateRepository::new(source.to_owned()));
-        let highlight: Box<liquid::Block> = Box::new(initialize_codeblock);
+        let highlight: Box<liquid::Block> =
+            Box::new(move |_, args, tokens, _| initialize_codeblock(args, tokens, &theme_name));
         options.blocks.insert("highlight".to_string(), highlight);
         let template = try!(liquid::parse(content, options));
         let html = try!(template.render(context)).unwrap_or_default();

--- a/src/document.rs
+++ b/src/document.rs
@@ -13,6 +13,7 @@ use jsonfeed::Item;
 use jsonfeed::Content;
 use serde_yaml;
 use itertools;
+use config::Config;
 
 #[cfg(all(feature="syntax-highlight", not(windows)))]
 use syntax_highlight::{initialize_codeblock, decorate_markdown};
@@ -379,7 +380,12 @@ impl Document {
     /// take `"extends"` attribute into account. This function can be used for
     /// rendering content or excerpt.
     #[cfg(all(feature="syntax-highlight", not(windows)))]
-    fn render_html(&self, content: &str, context: &mut Context, source: &Path) -> Result<String> {
+    fn render_html(&self,
+                   content: &str,
+                   context: &mut Context,
+                   source: &Path,
+                   config: &Config)
+                   -> Result<String> {
         let mut options = LiquidOptions::default();
         options.template_repository = Box::new(LocalTemplateRepository::new(source.to_owned()));
         let highlight: Box<liquid::Block> = Box::new(initialize_codeblock);
@@ -393,7 +399,7 @@ impl Document {
                 let mut buf = String::new();
                 let parser = cmark::Parser::new(&html);
                 #[cfg(feature="syntax-highlight")]
-                cmark::html::push_html(&mut buf, decorate_markdown(parser));
+                cmark::html::push_html(&mut buf, decorate_markdown(parser, config));
                 #[cfg(not(feature="syntax-highlight"))]
                 cmark::html::push_html(&mut buf, parser);
                 buf
@@ -402,7 +408,12 @@ impl Document {
         Ok(html.to_owned())
     }
     #[cfg(any(not(feature="syntax-highlight"), windows))]
-    fn render_html(&self, content: &str, context: &mut Context, source: &Path) -> Result<String> {
+    fn render_html(&self,
+                   content: &str,
+                   context: &mut Context,
+                   source: &Path,
+                   _: &Config)
+                   -> Result<String> {
         let mut options = LiquidOptions::default();
         options.template_repository = Box::new(LocalTemplateRepository::new(source.to_owned()));
         let template = liquid::parse(content, options)?;
@@ -439,7 +450,11 @@ impl Document {
     }
 
     /// Renders excerpt and adds it to attributes of the document.
-    pub fn render_excerpt(&mut self, context: &mut Context, source: &Path) -> Result<()> {
+    pub fn render_excerpt(&mut self,
+                          context: &mut Context,
+                          source: &Path,
+                          config: &Config)
+                          -> Result<()> {
         let excerpt_html = {
             let excerpt_attr = self.attributes
                 .get("excerpt")
@@ -448,13 +463,14 @@ impl Document {
             let excerpt_separator = &self.front.excerpt_separator;
 
             if let Some(excerpt_str) = excerpt_attr {
-                try!(self.render_html(excerpt_str, context, source))
+                try!(self.render_html(excerpt_str, context, source, config))
             } else if excerpt_separator.is_empty() {
-                try!(self.render_html("", context, source))
+                try!(self.render_html("", context, source, config))
             } else {
                 try!(self.render_html(&self.extract_markdown_references(excerpt_separator),
                                       context,
-                                      source))
+                                      source,
+                                      config))
             }
         };
 
@@ -476,9 +492,10 @@ impl Document {
                   context: &mut Context,
                   source: &Path,
                   layouts_dir: &Path,
-                  layouts_cache: &mut HashMap<String, String>)
+                  layouts_cache: &mut HashMap<String, String>,
+                  config: &Config)
                   -> Result<String> {
-        let content_html = try!(self.render_html(&self.content, context, source));
+        let content_html = try!(self.render_html(&self.content, context, source, config));
         self.attributes
             .insert("content".to_owned(), Value::Str(content_html.clone()));
         context.set_val("content", Value::Str(content_html.clone()));

--- a/src/error.rs
+++ b/src/error.rs
@@ -26,5 +26,10 @@ error_chain! {
             display("name, description and link need to be defined in the config file to \
                     generate RSS")
         }
+
+        UnsupportedPlatform(functionality: &'static str, platform: &'static str) {
+            description("functionality is not implemented for this platform")
+            display("{} is not implemented for the {} platform", functionality, platform)
+        }
     }
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -60,5 +60,5 @@ mod frontmatter;
 #[cfg(feature="syntax-highlight")]
 mod syntax_highlight;
 
-#[cfg(all(feature="syntax-highlight", not(windows)))]
+#[cfg(all(feature="syntax-highlight"))]
 pub use syntax_highlight::list_syntax_themes;

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -59,3 +59,6 @@ mod frontmatter;
 
 #[cfg(feature="syntax-highlight")]
 mod syntax_highlight;
+
+#[cfg(all(feature="syntax-highlight", not(windows)))]
+pub use syntax_highlight::list_syntax_themes;

--- a/src/main.rs
+++ b/src/main.rs
@@ -207,12 +207,8 @@ fn run() -> Result<()> {
                                  .value_name("COMMIT-MESSAGE")
                                  .help("Commit message that will be used on import")
                                  .default_value("cobalt site import")
-                                 .takes_value(true)));
-
-    #[cfg(all(feature="syntax-highlight"))]
-    let app_cli =
-        app_cli
-            .subcommand(SubCommand::with_name("list-syntax-themes").about("list available themes"));
+                                 .takes_value(true)))
+        .subcommand(SubCommand::with_name("list-syntax-themes").about("list available themes"));
 
     let global_matches = app_cli.get_matches();
 
@@ -418,7 +414,9 @@ fn run() -> Result<()> {
             }
 
             #[cfg(not(feature="syntax-highlight"))]
-            bail!(global_matches.usage());
+            bail!(concat!("No themes available.",
+                          "  This build of cobalt does not include",
+                          " support for syntax highlighting."));
         }
 
         _ => {

--- a/src/main.rs
+++ b/src/main.rs
@@ -209,7 +209,7 @@ fn run() -> Result<()> {
                                  .default_value("cobalt site import")
                                  .takes_value(true)));
 
-    #[cfg(all(feature="syntax-highlight", not(windows)))]
+    #[cfg(all(feature="syntax-highlight"))]
     let app_cli =
         app_cli
             .subcommand(SubCommand::with_name("list-syntax-themes").about("list available themes"));
@@ -412,12 +412,12 @@ fn run() -> Result<()> {
         }
 
         "list-syntax-themes" => {
-            #[cfg(all(feature="syntax-highlight", not(windows)))]
+            #[cfg(all(feature="syntax-highlight"))]
             for name in list_syntax_themes() {
                 println!("{}", name);
             }
 
-            #[cfg(any(not(feature="syntax-highlight"), windows))]
+            #[cfg(not(feature="syntax-highlight"))]
             bail!(global_matches.usage());
         }
 

--- a/src/main.rs
+++ b/src/main.rs
@@ -42,6 +42,9 @@ use std::io::prelude::*;
 use std::io::Result as IoResult;
 use std::fs::File;
 
+#[cfg(all(feature="syntax-highlight", not(windows)))]
+use cobalt::list_syntax_themes;
+
 error_chain! {
 
     links {
@@ -60,7 +63,7 @@ error_chain! {
 quick_main!(run);
 
 fn run() -> Result<()> {
-    let global_matches = App::new("Cobalt")
+    let app_cli = App::new("Cobalt")
         .version(crate_version!())
         .author("Benny Klotz <r3qnbenni@gmail.com>, Johann Hofmann")
         .about("A static site generator written in Rust.")
@@ -204,8 +207,14 @@ fn run() -> Result<()> {
                                  .value_name("COMMIT-MESSAGE")
                                  .help("Commit message that will be used on import")
                                  .default_value("cobalt site import")
-                                 .takes_value(true)))
-        .get_matches();
+                                 .takes_value(true)));
+
+    #[cfg(all(feature="syntax-highlight", not(windows)))]
+    let app_cli =
+        app_cli
+            .subcommand(SubCommand::with_name("list-syntax-themes").about("list available themes"));
+
+    let global_matches = app_cli.get_matches();
 
     let (command, matches) = match global_matches.subcommand() {
         (command, Some(matches)) => (command, matches),
@@ -400,6 +409,16 @@ fn run() -> Result<()> {
             let branch = matches.value_of("branch").unwrap().to_string();
             let message = matches.value_of("message").unwrap().to_string();
             import(&config, &branch, &message);
+        }
+
+        "list-syntax-themes" => {
+            #[cfg(all(feature="syntax-highlight", not(windows)))]
+            for name in list_syntax_themes() {
+                println!("{}", name);
+            }
+
+            #[cfg(any(not(feature="syntax-highlight"), windows))]
+            bail!(global_matches.usage());
         }
 
         _ => {

--- a/src/syntax_highlight.rs
+++ b/src/syntax_highlight.rs
@@ -150,7 +150,11 @@ pub fn has_syntax_theme(name: &str) -> bool {
 }
 
 pub fn list_syntax_themes<'a>() -> Vec<&'a String> {
-    SETUP.theme_set.themes.keys().collect::<Vec<_>>()
+    #[cfg(not(windows))]
+    return SETUP.theme_set.themes.keys().collect::<Vec<_>>();
+
+    #[cfg(windows)]
+    return vec![];
 }
 
 #[cfg(test)]

--- a/src/syntax_highlight.rs
+++ b/src/syntax_highlight.rs
@@ -10,7 +10,7 @@ use liquid::lexer::Element::{self, Expression, Tag, Raw};
 use liquid::Error;
 
 use syntect::parsing::SyntaxSet;
-use syntect::highlighting::ThemeSet;
+use syntect::highlighting::{ThemeSet, Theme};
 use syntect::html::{IncludeBackground, highlighted_snippet_for_string, styles_to_coloured_html,
                     start_coloured_html_snippet};
 use syntect::easy::HighlightLines;
@@ -20,8 +20,7 @@ use std::borrow::Cow::Owned;
 use self::cmark::Parser;
 use self::cmark::Tag as cmarkTag;
 use self::cmark::Event::{self, Start, End, Text, Html};
-
-const THEME_NAME: &'static str = "base16-ocean.dark";
+use config::Config;
 
 struct Setup {
     syntax_set: SyntaxSet,
@@ -41,11 +40,11 @@ lazy_static!{
 struct CodeBlock {
     lang: Option<String>,
     code: String,
+    theme: Theme,
 }
 
 impl Renderable for CodeBlock {
     fn render(&self, _: &mut Context) -> Result<Option<String>, Error> {
-
         let syntax = match self.lang {
                 Some(ref lang) => SETUP.syntax_set.find_syntax_by_token(lang),
                 _ => None,
@@ -54,20 +53,22 @@ impl Renderable for CodeBlock {
 
         Ok(Some(highlighted_snippet_for_string(&self.code,
                                                syntax,
-                                               &SETUP.theme_set.themes[THEME_NAME])))
+                                               &SETUP.theme_set.themes[&self.theme.name.unwrap()])))
     }
 }
 
 pub struct DecoratedParser<'a> {
     h: Option<HighlightLines<'a>>,
     parser: Parser<'a>,
+    config: &'a Config,
 }
 
 impl<'a> DecoratedParser<'a> {
-    pub fn new(parser: Parser<'a>) -> Self {
+    pub fn new(parser: Parser<'a>, config: &'a Config) -> Self {
         DecoratedParser {
             h: None,
             parser: parser,
+            config: config,
         }
     }
 }
@@ -95,10 +96,11 @@ impl<'a> Iterator for DecoratedParser<'a> {
                                 .next()
                                 .and_then(|lang| SETUP.syntax_set.find_syntax_by_token(lang))
                                 .unwrap_or_else(|| SETUP.syntax_set.find_syntax_plain_text());
-                        self.h = Some(HighlightLines::new(cur_syntax,
-                                                          &SETUP.theme_set.themes[THEME_NAME]));
+                        self.h = Some(HighlightLines::new(&cur_syntax,
+                                                          &SETUP.theme_set.themes
+                                                               [&self.config.theme]));
                         let snippet = start_coloured_html_snippet(&SETUP.theme_set.themes
-                                                                       [THEME_NAME]);
+                                                                       [&self.config.theme]);
                         return Some(Html(Owned(snippet)));
                     }
                     if let End(cmarkTag::CodeBlock(_)) = item {
@@ -119,7 +121,8 @@ impl<'a> Iterator for DecoratedParser<'a> {
 pub fn initialize_codeblock(_: &str,
                             arguments: &[Token],
                             tokens: &[Element],
-                            _: &LiquidOptions)
+                            _: &LiquidOptions,
+                            config: &Config)
                             -> Result<Box<Renderable>, Error> {
 
     let content = tokens
@@ -141,11 +144,12 @@ pub fn initialize_codeblock(_: &str,
     Ok(Box::new(CodeBlock {
                     code: content,
                     lang: lang,
+                    theme: config.theme,
                 }))
 }
 
-pub fn decorate_markdown(parser: Parser) -> DecoratedParser {
-    DecoratedParser::new(parser)
+pub fn decorate_markdown<'a>(parser: Parser<'a>, config: &'a Config) -> DecoratedParser<'a> {
+    DecoratedParser::new(parser, config)
 }
 
 #[cfg(test)]

--- a/src/syntax_highlight.rs
+++ b/src/syntax_highlight.rs
@@ -145,6 +145,10 @@ pub fn decorate_markdown<'a>(parser: Parser<'a>, config: &'a Config) -> Decorate
     DecoratedParser::new(parser, config)
 }
 
+pub fn has_syntax_theme(name: &str) -> bool {
+    SETUP.theme_set.themes.contains_key(name)
+}
+
 #[cfg(test)]
 mod test {
 

--- a/src/syntax_highlight.rs
+++ b/src/syntax_highlight.rs
@@ -98,9 +98,9 @@ impl<'a> Iterator for DecoratedParser<'a> {
                                 .unwrap_or_else(|| SETUP.syntax_set.find_syntax_plain_text());
                         self.h = Some(HighlightLines::new(&cur_syntax,
                                                           &SETUP.theme_set.themes
-                                                               [&self.config.theme]));
+                                                               [&self.config.syntax_theme]));
                         let snippet = start_coloured_html_snippet(&SETUP.theme_set.themes
-                                                                       [&self.config.theme]);
+                                                                       [&self.config.syntax_theme]);
                         return Some(Html(Owned(snippet)));
                     }
                     if let End(cmarkTag::CodeBlock(_)) = item {
@@ -144,7 +144,7 @@ pub fn initialize_codeblock(_: &str,
     Ok(Box::new(CodeBlock {
                     code: content,
                     lang: lang,
-                    theme: config.theme,
+                    theme: config.syntax_theme,
                 }))
 }
 

--- a/src/syntax_highlight.rs
+++ b/src/syntax_highlight.rs
@@ -149,6 +149,10 @@ pub fn has_syntax_theme(name: &str) -> bool {
     SETUP.theme_set.themes.contains_key(name)
 }
 
+pub fn list_syntax_themes<'a>() -> Vec<&'a String> {
+    SETUP.theme_set.themes.keys().collect::<Vec<_>>()
+}
+
 #[cfg(test)]
 mod test {
 

--- a/src/syntax_highlight.rs
+++ b/src/syntax_highlight.rs
@@ -2,6 +2,8 @@ extern crate syntect;
 extern crate liquid;
 extern crate pulldown_cmark as cmark;
 
+use error;
+
 use liquid::Renderable;
 use liquid::Context;
 use liquid::Token::{self, Identifier};
@@ -143,8 +145,15 @@ pub fn decorate_markdown<'a>(parser: Parser<'a>, theme_name: &str) -> DecoratedP
     DecoratedParser::new(parser, &SETUP.theme_set.themes[theme_name])
 }
 
-pub fn has_syntax_theme(name: &str) -> bool {
-    SETUP.theme_set.themes.contains_key(name)
+pub fn has_syntax_theme(name: &str) -> error::Result<bool> {
+    #[cfg(not(windows))]
+    return Ok(SETUP.theme_set.themes.contains_key(name));
+
+    #[cfg(windows)]
+    {
+        use error::ErrorKind;
+        return Err(ErrorKind::UnsupportedPlatform("syntax highlighting", "windows").into());
+    }
 }
 
 pub fn list_syntax_themes<'a>() -> Vec<&'a String> {

--- a/src/syntax_highlight.rs
+++ b/src/syntax_highlight.rs
@@ -4,7 +4,6 @@ extern crate pulldown_cmark as cmark;
 
 use liquid::Renderable;
 use liquid::Context;
-use liquid::LiquidOptions;
 use liquid::Token::{self, Identifier};
 use liquid::lexer::Element::{self, Expression, Tag, Raw};
 use liquid::Error;
@@ -51,9 +50,7 @@ impl Renderable for CodeBlock {
             }
             .unwrap_or_else(|| SETUP.syntax_set.find_syntax_plain_text());
 
-        Ok(Some(highlighted_snippet_for_string(&self.code,
-                                               syntax,
-                                               &SETUP.theme_set.themes[&self.theme.name.unwrap()])))
+        Ok(Some(highlighted_snippet_for_string(&self.code, syntax, &self.theme)))
     }
 }
 
@@ -96,11 +93,9 @@ impl<'a> Iterator for DecoratedParser<'a> {
                                 .next()
                                 .and_then(|lang| SETUP.syntax_set.find_syntax_by_token(lang))
                                 .unwrap_or_else(|| SETUP.syntax_set.find_syntax_plain_text());
-                        self.h = Some(HighlightLines::new(&cur_syntax,
-                                                          &SETUP.theme_set.themes
-                                                               [&self.config.syntax_theme]));
-                        let snippet = start_coloured_html_snippet(&SETUP.theme_set.themes
-                                                                       [&self.config.syntax_theme]);
+                        let theme = &SETUP.theme_set.themes[&self.config.syntax_theme];
+                        self.h = Some(HighlightLines::new(&cur_syntax, theme));
+                        let snippet = start_coloured_html_snippet(theme);
                         return Some(Html(Owned(snippet)));
                     }
                     if let End(cmarkTag::CodeBlock(_)) = item {
@@ -118,11 +113,9 @@ impl<'a> Iterator for DecoratedParser<'a> {
     }
 }
 
-pub fn initialize_codeblock(_: &str,
-                            arguments: &[Token],
+pub fn initialize_codeblock(arguments: &[Token],
                             tokens: &[Element],
-                            _: &LiquidOptions,
-                            config: &Config)
+                            theme_name: &str)
                             -> Result<Box<Renderable>, Error> {
 
     let content = tokens
@@ -144,7 +137,7 @@ pub fn initialize_codeblock(_: &str,
     Ok(Box::new(CodeBlock {
                     code: content,
                     lang: lang,
-                    theme: config.syntax_theme,
+                    theme: SETUP.theme_set.themes[theme_name].clone(),
                 }))
 }
 
@@ -188,7 +181,10 @@ mod test {
         let mut options: LiquidOptions = Default::default();
         options
             .blocks
-            .insert("codeblock".to_string(), Box::new(initialize_codeblock));
+            .insert("codeblock".to_string(),
+                    Box::new(|_, args, tokens, _| {
+                                 initialize_codeblock(args, tokens, "base16-ocean.dark")
+                             }));
         let template = liquid::parse(&format!("{{% codeblock rust %}}{}{{% endcodeblock %}}",
                                               CODE_BLOCK),
                                      options)

--- a/src/syntax_highlight.rs
+++ b/src/syntax_highlight.rs
@@ -19,7 +19,6 @@ use std::borrow::Cow::Owned;
 use self::cmark::Parser;
 use self::cmark::Tag as cmarkTag;
 use self::cmark::Event::{self, Start, End, Text, Html};
-use config::Config;
 
 struct Setup {
     syntax_set: SyntaxSet,
@@ -57,15 +56,15 @@ impl Renderable for CodeBlock {
 pub struct DecoratedParser<'a> {
     h: Option<HighlightLines<'a>>,
     parser: Parser<'a>,
-    config: &'a Config,
+    theme: &'a Theme,
 }
 
 impl<'a> DecoratedParser<'a> {
-    pub fn new(parser: Parser<'a>, config: &'a Config) -> Self {
+    pub fn new(parser: Parser<'a>, theme: &'a Theme) -> Self {
         DecoratedParser {
             h: None,
             parser: parser,
-            config: config,
+            theme: &theme,
         }
     }
 }
@@ -93,9 +92,8 @@ impl<'a> Iterator for DecoratedParser<'a> {
                                 .next()
                                 .and_then(|lang| SETUP.syntax_set.find_syntax_by_token(lang))
                                 .unwrap_or_else(|| SETUP.syntax_set.find_syntax_plain_text());
-                        let theme = &SETUP.theme_set.themes[&self.config.syntax_theme];
-                        self.h = Some(HighlightLines::new(&cur_syntax, theme));
-                        let snippet = start_coloured_html_snippet(theme);
+                        self.h = Some(HighlightLines::new(&cur_syntax, self.theme));
+                        let snippet = start_coloured_html_snippet(self.theme);
                         return Some(Html(Owned(snippet)));
                     }
                     if let End(cmarkTag::CodeBlock(_)) = item {
@@ -141,8 +139,8 @@ pub fn initialize_codeblock(arguments: &[Token],
                 }))
 }
 
-pub fn decorate_markdown<'a>(parser: Parser<'a>, config: &'a Config) -> DecoratedParser<'a> {
-    DecoratedParser::new(parser, config)
+pub fn decorate_markdown<'a>(parser: Parser<'a>, theme_name: &str) -> DecoratedParser<'a> {
+    DecoratedParser::new(parser, &SETUP.theme_set.themes[theme_name])
 }
 
 pub fn has_syntax_theme(name: &str) -> bool {

